File: 09-backtracking/00040-Combination_Sum_II/00040-combination_sum_II.cpp
// 40: Combination Sum Ii
// https://leetcode.com/problems/combination-sum-ii


#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
private:
    void backtrack(vector<int>& candidates, int target, int sum, int start, vector<int>& current, vector<vector<int>>& result) {
        if (sum > target) return;

        if (sum == target) {
            result.push_back(current);
            return;
        }

        for (int i=start; i<candidates.size(); i++) {
            if (i>start && candidates[i] == candidates[i-1]) continue;
            current.push_back(candidates[i]);
            backtrack(candidates, target, sum + candidates[i], i+1, current, result);
            current.pop_back();
        }
    }

public:
    // SOLUTOIN
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        std::sort(candidates.begin(), candidates.end());
        vector<int> current;
        vector<vector<int>> result;
        backtrack(candidates, target, 0, 0, current, result);
        return result;
    }
};


int main() {
    Solution o;

    // INPUT
    vector<int> candidates = {2,5,2,1,2};
    int target = 5;

    // OUTPUT
    auto result = o.combinationSum2(candidates, target);

    cout << "["; for (auto x : result) {
        cout << "["; for (auto y : x) {
            cout << y << ",";
        } cout << "\b],";
    } cout << "\b]" << endl;

    return 0;
}

File: 09-backtracking/00040-Combination_Sum_II/00040-combination_sum_II.java
// 40: Combination Sum Ii
// https://leetcode.com/problems/combination-sum-ii


import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;

class Solution {
    private void backtrack(int[] candidates, int target, int sum, int start, List<Integer> current, List<List<Integer>> result) {
        if (sum > target) return;

        if (sum == target) {
            result.add(new ArrayList<Integer>(current));
            return;
        }

        for (int i=start; i<candidates.length; i++) {
            if (i>start && candidates[i] == candidates[i-1]) continue;
            current.add(candidates[i]);
            backtrack(candidates, target, sum + candidates[i], i+1, current, result);
            current.remove(current.size()-1);
        }
    }

    // SOLUTION
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> result = new ArrayList<>();
        List<Integer> current = new ArrayList<>();
        backtrack(candidates, target, 0, 0, current, result);
        return result;
    }

    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        int[] candidates = {2,5,2,1,2};
        int target = 5;

        // OUTPUT
        var result = o.combinationSum2(candidates, target);
        System.out.println(result);
    }
}
